### PR TITLE
Add generic `Shoot` `ConfigMap` validation webhook handler

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area dev-productivity
/kind enhancement

**What this PR does / why we need it**:
Common parts from the following handlers have been moved here:
- https://github.com/gardener/gardener/blob/master/pkg/admissioncontroller/webhook/admission/auditpolicy/handler.go
- https://github.com/gardener/gardener/blob/master/pkg/admissioncontroller/webhook/admission/authenticationconfig/handler.go
- https://github.com/gardener/gardener/blob/master/pkg/admissioncontroller/webhook/admission/authorizationconfig/handler.go

This drastically reduces duplicating the same boilerplate again and again.

**Which issue(s) this PR fixes**:
Follow-up of #10682 

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: pkg/webhook/configvalidator/handler.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+}
+
+// Handler validates configuration part of ConfigMaps which are referenced in Shoot resources.
+type Handler struct {
+	Logger    logr.Logger
+	APIReader client.Reader
+	Client    client.Reader
+	Decoder   admission.Decoder
+
+	ConfigMapKind               string
```
Comment (plkokanov) [inline pkg/webhook/configvalidator/handler.go]: Not too strong of an opinion, but the `Kind` suffix is a bit confusing as it is also used to refer to the actual k8s resource `Kind` field. What about using `Purpose`?
```suggestion
	ConfigMapPurpose               string
```

File: pkg/admissioncontroller/webhook/admission/authenticationconfig/handler.go [hunk 2]
```diff
@@ -1,244 +0,0 @@
...
-
-	for _, shoot := range shootList.Items {
-		if v1beta1helper.GetShootAuthenticationConfigurationConfigMapName(shoot.Spec.Kubernetes.KubeAPIServer) == request.Name {
-			configMapIsReferenced = true
-			coreKubeAPIServerConfig := &gardencore.KubeAPIServerConfig{}
-			if err := gardencorev1beta1.Convert_v1beta1_KubeAPIServerConfig_To_core_KubeAPIServerConfig(shoot.Spec.Kubernetes.KubeAPIServer, coreKubeAPIServerConfig, nil); err != nil {
-				return admission.Errored(http.StatusUnprocessableEntity, err)
-			}
-			disallowedIssuers = getDisallowedIssuers(coreKubeAPIServerConfig)
-			break
```
Comment (plkokanov) [inline pkg/admissioncontroller/webhook/admission/authenticationconfig/handler.go:174]: We used to break on the first encountered shoot that references the given ConfigMap and only `getDisallowedIssuers` from it. 

Now we will get the disallowed issuers from all shoots that reference the ConfigMap and use them for the validation which, I guess, improves the validation 🎉:  https://github.com/gardener/gardener/blob/2e687f3b129071f5cb443d1535fc7fab3a4ea59a/pkg/admissioncontroller/webhook/admission/authenticationconfig/add.go#L100-L106

File: pkg/admissioncontroller/webhook/admission/auditpolicy/add_test.go [hunk 3]
```diff
@@ -178,27 +178,27 @@ rules:
 
 		if oldObj != nil {
 			objData, err := runtime.Encode(testEncoder, oldObj)
-			Expect(err).NotTo(HaveOccurred())
+			ExpectWithOffset(1, err).NotTo(HaveOccurred())
```
Comment (marc1404) [inline pkg/admissioncontroller/webhook/admission/auditpolicy/add_test.go:181]: Good catch 🦅

**Review by marc1404 — COMMENTED** [review]
Thanks for reducing the boilerplate repetition! 🧹 
+1 for improving the error messages and tests overall :) 

/lgtm

File: pkg/admissioncontroller/webhook/admission/authorizationconfig/add.go [hunk 4]
```diff
@@ -17,13 +37,87 @@ const (
...
+			if !apiequality.Semantic.DeepEqual(webhook.Webhook.ConnectionInfo, apiserver.WebhookConnectionInfo{}) {
+				return http.StatusUnprocessableEntity, fmt.Errorf("provided invalid authorization configuration: %v", field.ErrorList{field.Forbidden(field.NewPath("authorizers").Index(i).Child("connectionInfo"), "connectionInfo is not allowed to be set")})
+			}
+			// At this point, we have ensured that the connection info is not set. However, below validation function
+			// expects it to be set, so let's fake it here to make the function pass (this is not persisted anywhere).
+			authorizationConfig.Authorizers[i].Webhook.ConnectionInfo.Type = apiserver.AuthorizationWebhookConnectionInfoTypeInCluster
+		}
+	}
+
+	if len(shoots) == 1 {
```
Comment (plkokanov) [inline pkg/admissioncontroller/webhook/admission/authorizationconfig/add.go]: Isn't it more correct to validate the `authorizerNames` against all of the shoots that reference the `ConfigMap` (even though this wasn't done before)?

I'm imagining the following case  that could happen currently:
1. 2 shoots have enabled structured authorization. 
2. Each of them has 1 entry in `.spec.kubernetes.kubeAPIServer.structuredAuthorization.kubeconfigs[]`. 
3. An operator adds an additional entry to that array for the first shoot. 
4. The operator modifies the referenced structured authorization ConfigMap to add an additional webhook that would match the new kubeconfig entry which is only available in one of the shoots
5. Currently, the validation passes as the length of shoots is > 1

**Review by plkokanov — COMMENTED** [review]
Nice refactoring! Just one small suggestion and a question inline.

File: pkg/admissioncontroller/webhook/admission/authenticationconfig/handler.go (hunk 2 above)
Comment (rfranzke) [reply pkg/admissioncontroller/webhook/admission/authenticationconfig/handler.go:174]: Yes, I don't know why there was this `break`. This was wrong/a bug FMPOV, so I removed this.

File: pkg/admissioncontroller/webhook/admission/authorizationconfig/add.go (hunk 4 above)
Comment (rfranzke) [reply pkg/admissioncontroller/webhook/admission/authorizationconfig/add.go]: Yes, definitely, good catch. Unfortunately, I don't remember the exact reason why it wasn't done this way in the initial implementation, but it was a deliberate decision. I think it didn't fit quite well into the validator structure copied from the other existing validators at this time. Anyways, I added it now - I think it's pretty straight-forward now with the new approach :) PTAL

File: pkg/webhook/configvalidator/handler.go [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+	}
+
+	// `oldConfigMapName` is empty for CREATE shoot requests that specify a ConfigMap reference, hence we can skip the
+	// validation.
+	// Additionally, we only need to revalidate if the ConfigMap data is compatible with a new Kubernetes version if the
+	// version changed.
+	if oldShoot != nil &&
+		h.GetConfigMapNameFromShoot(oldShoot) == configMapName &&
+		oldShoot.Spec.Kubernetes.Version == shoot.Spec.Kubernetes.Version &&
+		(h.SkipValidationOnShootUpdate == nil || h.SkipValidationOnShootUpdate(shoot, oldShoot)) {
```
Comment (plkokanov) [inline pkg/webhook/configvalidator/handler.go:110]: On a second read, I think we skip some validation cases here that we wouldn't want to skip. For instance, the `authorizationconfig` handler does not set a `h.SkipValidationOnShootUpdate` function, so if the ConfigMap name didn't change and the k8s version didn't change, but the `shoot.spec.kubernetes.kubeAPIServer.StructuredAuthorization.Kubeconfigs` changed, the validation would still be skipped.


I guess, `h.SkipValidationOnShootUpdate` should be set for the `authorizationconfig` handler, similar to the `authenticationconfig` one.

---

Alternatively, as a personal preference, I think it makes a bit more sense to not skip the validation if `h.SkipValidationOnShootUpdate` is not set, so what about flipping this check (bad thing is that we will then have to add a dummy function that always returns true for the audit policy handler):
```suggestion
		(h.SkipValidationOnShootUpdate != nil && h.SkipValidationOnShootUpdate(shoot, oldShoot)) {
```

**Review by plkokanov — CHANGES_REQUESTED** [review]

File: pkg/webhook/configvalidator/handler.go (hunk 5 above)
Comment (rfranzke) [reply pkg/webhook/configvalidator/handler.go:110]: > On a second read, I think we skip some validation cases here that we wouldn't want to skip. For instance, the `authorizationconfig` handler does not set a `h.SkipValidationOnShootUpdate` function, so if the ConfigMap name didn't change and the k8s version didn't change, but the `shoot.spec.kubernetes.kubeAPIServer.StructuredAuthorization.Kubeconfigs` changed, the validation would still be skipped.
> 
> I guess, `h.SkipValidationOnShootUpdate` should be set for the `authorizationconfig` handler, similar to the `authenticationconfig` one.

Good suggestion, thank you! :) Done!

> Alternatively, as a personal preference, I think it makes a bit more sense to not skip the validation if `h.SkipValidationOnShootUpdate` is not set, so what about flipping this check (bad thing is that we will then have to add a dummy function that always returns true for the audit policy handler):

I don't want to specify this function when not needed - that's why I prefer the current approach. We still only skip both `ConfigMap` name and Kubernetes version have not changed. Either the function is not set, or it is set and tells us to still skip. I don't understand why you prefer it the other way round with a dummy function 😕

Comment (plkokanov) [reply pkg/webhook/configvalidator/handler.go:110]: > I don't understand why you prefer it the other way round with a dummy function

I was making the analogy that the default value for a boolean is false -> this is a function that returns a boolean, so if the function itself is not set, the default behaviour should also be "false". But from a practical PoV you're right that having to add a dummy function is not too good, so I'm ok with keeping it as it is :)

File: pkg/admissioncontroller/webhook/admission/authorizationconfig/add_test.go [hunk 6]
```diff
@@ -237,6 +237,19 @@ authorizers:
...
+				returnedCm := corev1.ConfigMap{
+					Data: map[string]string{"config.yaml": validAuthorizationConfiguration},
+				}
+				newShoot := shoot.DeepCopy()
+				newShoot.Spec.Kubernetes.KubeAPIServer.StructuredAuthorization.Kubeconfigs[0].AuthorizerName = "does-not-exist"
+				mockReader.EXPECT().Get(gomock.Any(), client.ObjectKey{Namespace: shootNamespace, Name: configMapName}, gomock.AssignableToTypeOf(&corev1.ConfigMap{})).DoAndReturn(func(_ context.Context, _ client.ObjectKey, cm *corev1.ConfigMap, _ ...client.GetOption) error {
+					*cm = returnedCm
+					return nil
+				})
+				test(admissionv1.Update, shoot, newShoot, true, statusCodeAllowed, "provided invalid authorization configuration: must provide kubeconfig secret name reference for webhook authorizer \"webhook\" in shoot fake-cm-namespace/fake-shoot-name", "")
```
Comment (plkokanov) [inline pkg/admissioncontroller/webhook/admission/authorizationconfig/add_test.go]:
```suggestion
				test(admissionv1.Update, shoot, newShoot, false, statusCodeInvalid, "provided invalid authorization configuration: must provide kubeconfig secret name reference for webhook authorizer \"webhook\" in shoot fake-cm-namespace/fake-shoot-name", "")
```

And should be moved to the `Deny` context

**Review by plkokanov — CHANGES_REQUESTED** [review]

File: pkg/webhook/configvalidator/handler.go (hunk 5 above)
Comment (rfranzke) [reply pkg/webhook/configvalidator/handler.go:110]: OK, great. Thank you :)

File: pkg/admissioncontroller/webhook/admission/authorizationconfig/add_test.go (hunk 6 above)
Comment (rfranzke) [reply pkg/admissioncontroller/webhook/admission/authorizationconfig/add_test.go]: Done, sorry. Pushed too fast 🚤 :)

**Pull request merged**